addons/product/static/src/js/product_configurator/product_configurator_dialog/product_configurator_dialog.js

```diff
@@ -0,0 +1,404 @@
+        });
+
+        useSubEnv({
+            mainProductTmplId: this.props.productTemplateId,
+            setQuantity: this._setQuantity.bind(this),
+            updateProductTemplateSelectedPTAV: this._updateProductTemplateSelectedPTAV.bind(this),
+            updatePTAVCustomValue: this._updatePTAVCustomValue.bind(this),
+            isPossibleCombination: this._isPossibleCombination,
+            options: {
+                no_open: this.props.no_open,
+                showCustom: this.props.showCustom,
+                showQty: this.props.showQty,
+                qtyFieldName: this.props.qtyFieldName,
+                productUomIdFieldName: this.props.productUomIdFieldName
+            }
+        });
+
+        onWillStart(async () => {
+            await this._setProducts();
+            for (const customValue of this.props.customAttributeValues) {
+                this._updatePTAVCustomValue(
+                    this.env.mainProductTmplId,
+                    customValue.ptavId,
+                    customValue.value
+                );
+            }
+            await this._onWillStartCustomCheck();
+            this._checkExclusions(this.state.products[0]);
+        });
+    }
+
+    async _onWillStartCustomCheck(){
+        if(this.props.options?.showCustom){
+            return
+        }
+        for (let product of this.state.products){
+            for (let attribute_line of product.attribute_lines){
+                const firstNotCustom = attribute_line.attribute_values.filter(
+                    val => !val.is_custom);
+                if(firstNotCustom.length > 0){
+                    await this._updateProductTemplateSelectedPTAV(product.product_tmpl_id, attribute_line.id, firstNotCustom[0].id);
+                }
+            }
+        }
+    }
+
+    //--------------------------------------------------------------------------
+    // Data Exchanges
+    //--------------------------------------------------------------------------
+
+    async _loadData(onlyMainProduct) {
+        let params = {
+            product_template_id: this.props.productTemplateId,
+            company_id: this.props.companyId,
+            ptav_ids: this.props.ptavIds,
+            remove_custom: !this.props.options?.showCustom
+        };
+        if (this.props.options.showQty) {
+            params['quantity'] = this.props.quantity;
+            params['product_uom_id'] = this.props.productUOMId;
+        }
+
+        return this.rpc('/product_configurator/get_values', {...params, ...this._loadDataVals(onlyMainProduct)});
+    }
+
+    async _createProduct(product) {
+        return this.rpc('/product_configurator/create_product', {
+            product_template_id: product.product_tmpl_id,
+            combination: this._getCombination(product),
+        });
+    }
+
+    async _updateCombination(product, quantity) {
+        return this.rpc('/product_configurator/update_combination', this._updateCombinationVals(product, quantity));
```

Comment:
Same comments as I made before about `_loadData`. Please apply the comments everywhere it suits.
https://github.com/odoo/odoo/pull/129118#discussion_r1343708306